tests/handlers/test_register.py

```diff
@@ -127,7 +132,15 @@ def make_homeserver(self, reactor, clock):
         hs_config["max_mau_value"] = 50
         hs_config["limit_usage_by_mau"] = True
 
-        hs = self.setup_test_homeserver(config=hs_config)
+        # Don't attempt to reach out over federation.
+        self.mock_federation_client = Mock()
+        self.mock_federation_client.make_query.side_effect = CodeMessageException(
+            500, ""
+        )
```

Comment:
this will *return* an exception object, rather than raise it? Is that correct?

Reply:
Setting `side_effect` to an exception class or instance raises it.

From [the docs](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect):

> This can either be a function to be called when the mock is called, an iterable or an exception (class or instance) to be raised.
>
> ...
>
> An example of a mock that raises an exception (to test exception handling of an API):
>
> ```pycon
> >>> mock = Mock()
> >>> mock.side_effect = Exception('Boom!')
> >>> mock()
> Traceback (most recent call last):
>   ...
> Exception: Boom!
> ```

The `side_effect` property has a lot of confusing special behavior...

Reply:
ugh.